yarn-project/simulator/src/avm/journal/journal.ts

```diff
@@ -238,26 +258,51 @@ export class AvmPersistableStateManager {
    * @returns exists - whether the nullifier exists in the nullifier set
    */
   public async checkNullifierExists(contractAddress: Fr, nullifier: Fr): Promise<boolean> {
-    const [exists, isPending, leafIndex] = await this.nullifiers.checkExists(contractAddress, nullifier);
-    this.log.debug(
-      `nullifiers(${contractAddress})@${nullifier} ?? leafIndex: ${leafIndex}, exists: ${exists}, pending: ${isPending}.`,
-    );
+    const [exists, isPending, _] = await this.nullifiers.checkExists(contractAddress, nullifier);
+
     const siloedNullifier = siloNullifier(contractAddress, nullifier);
-    const lowResult = await this.merkleTree.getPreviousValueIndex(
-      MerkleTreeId.NULLIFIER_TREE,
-      siloedNullifier.toBigInt(),
-    );
-    const indexedLeafPreimage = await this.merkleTree.getLeafPreimage(MerkleTreeId.NULLIFIER_TREE, lowResult!.index);
-    const lowLeafPreimage = NullifierLeafPreimage.fromBuffer(indexedLeafPreimage!.toBuffer());
-    const lowLeafPath = await this.merkleTree.getSiblingPath(MerkleTreeId.NULLIFIER_TREE, lowResult!.index);
-    this.trace.traceNullifierCheck(
-      contractAddress,
-      nullifier, // Maybe this should be unsiloed
-      exists,
-      lowLeafPreimage,
-      new Fr(lowResult!.index),
-      lowLeafPath.toFields(),
-    );
+
+    if (this.doMerkleOperations) {
+      // Get leaf if present, low leaf if absent
+      // If leaf is present, hint/trace it. Otherwise, hint/trace the low leaf.
+      const [leafIndex, leafPreimage, leafPath, alreadyPresent] = await getLeafOrLowLeaf<NullifierLeafPreimage>(
+        MerkleTreeId.NULLIFIER_TREE,
+        siloedNullifier.toBigInt(),
+        this.merkleTrees,
```

Comment:
Pulled this into the helper function which gets the leafIndex, preimage, path, alreadyPresent. The leaf that it gets is the low-leaf unless the target nullifier actually exists in which case we get that leaf and trace it. Does that make sense? Or should we always get the low-nullifier here? `getPreviousValueIndex` only returns the low-leaf if the target leaf doesn't exist, hence why I went with this option. But I'm fine changing it back!